services/middlewareServices/FormSGProcessingService.ts

```diff
@@ -0,0 +1,74 @@
+import {
+  DecryptParams,
+  DecryptedContent,
+} from "@opengovsg/formsg-sdk/dist/types"
+import { Request, Response, NextFunction } from "express"
+
+import logger from "@logger/logger"
+
+import { AuthError } from "@root/errors/AuthError"
+
+export interface CanDecryptFormSGPayload {
+  webhooks: {
+    authenticate: (header: string, uri: string) => void
+  }
+  crypto: {
+    decrypt: (
+      formCreateKey: string,
+      decryptParams: DecryptParams
+    ) => DecryptedContent | null
+  }
+}
+export default class FormSGProcessingService {
+  formsg: CanDecryptFormSGPayload
+
+  constructor({ formsg }: { formsg: CanDecryptFormSGPayload }) {
+    this.formsg = formsg
+  }
+
+  authenticate = () => (
+    req: Request,
+    _res: Response,
+    next: NextFunction
+  ): void => {
+    const signature = req.get("X-FormSG-Signature")
+    if (!signature) {
+      throw new AuthError("Signature missing")
+    }
+    const postUri = `https://${req.get("host")}${req.baseUrl}${req.path}`
+    try {
+      this.formsg.webhooks.authenticate(signature, postUri)
+      // Continue processing the POST body
+      next()
+    } catch (e) {
+      logger.error(e)
+      throw new AuthError("Unauthorized")
+    }
+  }
+
+  decrypt = ({ formKey }: { formKey: string }) => (
+    req: Request,
+    res: Response,
+    next: NextFunction
+  ): void => {
+    try {
+      const submission = this.formsg.crypto.decrypt(
+        formKey,
+        // If `verifiedContent` is provided in `req.body.data`, the return object
+        // will include a verified key.
+        req.body.data
+      )
+
+      // If the decryption failed, submission will be `null`.
+      if (!submission) {
+        res.status(422).send({ message: "Bad submission" })
+      }
+      // Continue processing the submission
+      res.locals.submission = submission
+      next()
+    } catch (e) {
+      logger.error(e)
+      throw new AuthError("Unauthorized")
```

Comment:
just noticed that both errors returning from `decrypt` and `authenticate` are identical - we should aim to distinguish our errors and attach more meaningful information to aid in troubleshooting.

this could be done through extending the `AuthError` to `DecryptionError` and also attach more meta data to it (eg: `formKey`)

Reply:
I've updated the error messages to include the post URI and additional info in 344cfaa and d8909a7 - but we shouldn't be adding the `formKey` data, since that's supposed to be a secret key

Reply:
sure, that sounds good! didn't realise that `formKey` is secret oops, thanks for letting me know!